### PR TITLE
fix: warn if vLLM is set and OS is not Linux

### PR DESCRIPTION
**Pull request by leseb**
Running `ilab model serve --backend vllm` on macOS will never work since
vLLM is only supported on Linux.

The CLI now warns with:

```
$ ilab model serve --backend vllm
...
WARNING 2024-07-11 09:30:34,816 backends.py:195: get The serving backend 'vllm' was configured explicitly, but the provided model is not compatible with it. The model was detected as 'llama-cpp, reason: model is a GGUF file.'.
The backend startup sequence will continue with the configured backend but might fail.
```

And the backend initialization will fail later on.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
I'll fix the test failure once https://github.com/instructlab/instructlab/pull/1561 merges since they are related.

File: src/instructlab/model/backends/backends.py
Comment (tiran) [inline src/instructlab/model/backends/backends.py:116]: Please use a modul-level logger instead of passing loggers around. We got mostly rid of this bad practice a short while ago.

**Comment by alimaredia** [comment]
@leseb I understand having a `--force` flag to override auto-detection might be a best practice, but I can't think of a case today where it would be used. Did you have one you've hit already or were thinking of when you came up with this idea?

Also could you change the name of the PR to mention the `--force` flag you're adding.

File: src/instructlab/model/backends/backends.py
Comment (leseb) [reply src/instructlab/model/backends/backends.py:116]: Ah right, forgot about that, thanks for pointing it out.

**Comment by leseb** [comment]
> @leseb I understand having a `--force` flag to override auto-detection might be a best practice, but I can't think of a case today where it would be used. Did you have one you've hit already or were thinking of when you came up with this idea? I worry about introducing a flag that would never be used.
> 
> Also could you change the name of the PR to mention the `--force` flag you're adding.

I've encountered that issue as well, and it was mainly due to a bug. I see the flag as a safeguard to prevent someone from being blocked by a bug in auto-detection. It's a common practice to have the ability to override auto-detection with a flag. My intention was to fail if someone uses `--backend` and the auto-detection detected something else. However, we could assume that people using `--backend` or setting `backend: foo` know what they are doing and let the backend start failing later. I was simply trying to catch the problem early.

For now, I'll extract `--force` from this PR and discuss it further later.

Thanks!

File: src/instructlab/model/backends/backends.py
Comment (leseb) [reply src/instructlab/model/backends/backends.py:116]: PTAL https://github.com/instructlab/instructlab/pull/1604

**Comment by leseb** [comment]
This is now on top of https://github.com/instructlab/instructlab/pull/1604

**Review by alimaredia — DISMISSED** [review]

**Review by alimaredia — APPROVED** [review]

File: src/instructlab/model/backends/backends.py
Comment (russellb) [inline src/instructlab/model/backends/backends.py:178]:
```suggestion
                "The backend startup sequence will continue with the configured backend but might fail."
```

**Review by russellb — COMMENTED** [review]
I think there's a typo in a log message.

Otherwise, this seems like an improvement, though I was expecting to see a more explicit check and log message to point out "you configured vllm, but this isn't linux".  That failure mode doesn't seem obvious.  It's just going to say "you chose vllm, but we detected llama-cpp" but doesn't really explain why

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [inline src/instructlab/model/backends/backends.py:146]: should this be more explicit and say "..a directory containing huggingface safetensors files..." instead of just "tensor files" ?

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [inline src/instructlab/model/backends/backends.py]: This wording seems confusing to me, would it better to rephrase this as "The serving backend '{backend}' was requested explicitly, but the provided model is not compatible with it"?

**Comment by leseb** [comment]
> I think there's a typo in a log message.
> 
> Otherwise, this seems like an improvement, though I was expecting to see a more explicit check and log message to point out "you configured vllm, but this isn't linux". That failure mode doesn't seem obvious. It's just going to say "you chose vllm, but we detected llama-cpp" but doesn't really explain why

Good point how about this new approach? Thanks!

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [inline src/instructlab/model/backends/backends.py:131]: perhaps we should be checking for the existence of .safetensors files instead of just checking if the supplied model is a directory
this could be fixed in a follow up PR though

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [inline src/instructlab/model/backends/backends.py:179]: This wording seems confusing to me, could we considering rephrasing this to something along the lines of "The serving backend '{backend}' was requested explicity, but the provided model is not compatible with it"?

**Review by jaideepr97 — COMMENTED** [review]
Thanks for the PR @leseb! Had a couple thoughts

**Review by russellb — COMMENTED** [review]
I'm good with this pending @jaideepr97's suggestions for wording changes

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [reply src/instructlab/model/backends/backends.py:131]: nvm, @alimaredia pointed me to https://github.com/instructlab/instructlab/issues/1457

Comment (leseb) [reply src/instructlab/model/backends/backends.py:131]: Today, we don't check for the tensors files, just for a directory, there is an improvement filed here https://github.com/instructlab/instructlab/issues/1457.

However I'll make it clearer.

**Comment by leseb** [comment]
@jaideepr97 wording updated, PTAL!

File: tests/test_backends.py
Comment (tiran) [inline tests/test_backends.py:25]: Use pytest fixture for temporary files:

```suggestion
def test_get_backend_auto_detection_fail_not_gguf(tmp_path: pathlib.Path):
    tmp_gguf = tmp_path / "test.gguf"
```

File: tests/test_backends.py
Comment (tiran) [inline tests/test_backends.py:26]: Random is problematic in tests and make them nondeterministic. This code has a non-zero change to have a valid GGUF header.

File: tests/test_backends.py
Comment (tiran) [inline tests/test_backends.py:45]:
```suggestion
@patch("instructlab.model.backends.backends.is_model_gguf", return_value=True)
def test_get_backend_auto_detection_success_gguf(m_is_model_gguf, tmp_path: pathlib.Path):
    tmp_gguf = tmp_path / "test.gguf"
    backend = backends.get(file, None)
    assert backend == "llama-cpp"
    m_is_model_gguf.assert_called_once_with(...)
```

**Review by tiran — COMMENTED** [review]
I recommend to use `tmp_path` fixture and `@patch` decorator.

File: tests/test_backends.py
Comment (leseb) [reply tests/test_backends.py]: Valid concern addressed, thanks!

**Review by tiran — APPROVED** [review]

**Review by russellb — APPROVED** [review]
thanks for the updates

**Review by jaideepr97 — APPROVED** [review]
LGTM, thanks @leseb

**Comment by leseb** [comment]
> @leseb since @tiran already gave you an approval before your last force push, feel free to use your judgement on whether you want him to re-review. If you don't think he needs to re-review feel free to dismiss his pending review.

I already did request @tiran to review again, I don't want to dismiss his review, I want a final review.

**Review by tiran — APPROVED** [review]

**Pull request merged**
